llvm/lib/Target/DirectX/DXILLegalizePass.cpp

```diff
@@ -151,6 +152,82 @@ downcastI64toI32InsertExtractElements(Instruction &I,
   }
 }
 
+void emitMemset(IRBuilder<> &Builder, Value *Dst, Value *Val,
+                ConstantInt *SizeCI) {
+  LLVMContext &Ctx = Builder.getContext();
+  [[maybe_unused]] DataLayout DL =
+      Builder.GetInsertBlock()->getModule()->getDataLayout();
+  [[maybe_unused]] uint64_t OrigSize = SizeCI->getZExtValue();
+
+  AllocaInst *Alloca = dyn_cast<AllocaInst>(Dst);
+
+  assert(Alloca && "Expected memset on an Alloca");
+  assert(OrigSize == Alloca->getAllocationSize(DL)->getFixedValue() &&
+         "Expected for memset size to match DataLayout size");
+
+  Type *AllocatedTy = Alloca->getAllocatedType();
+  ArrayType *ArrTy = dyn_cast<ArrayType>(AllocatedTy);
+  assert(ArrTy && "Expected Alloca for an Array Type");
+
+  Type *ElemTy = ArrTy->getElementType();
+  uint64_t Size = ArrTy->getArrayNumElements();
+
+  [[maybe_unused]] uint64_t ElemSize = DL.getTypeStoreSize(ElemTy);
+
+  assert(ElemSize > 0 && "Size must be set");
+  assert(OrigSize == ElemSize * Size && "Size in bytes must match");
+
+  Value *TypedVal = Val;
+  if (Val->getType() != ElemTy)
+    TypedVal = Builder.CreateIntCast(Val, ElemTy,
+                                     false); // Or use CreateBitCast for float
+
+  for (uint64_t I = 0; I < Size; ++I) {
+    Value *Offset = ConstantInt::get(Type::getInt32Ty(Ctx), I);
+    Value *Ptr = Builder.CreateGEP(ElemTy, Dst, Offset, "gep");
+    Builder.CreateStore(TypedVal, Ptr);
+  }
+}
+
+void removeLifetimesForMemset(CallInst *Memset,
```

Comment:
I'm not sure we should be deleting lifetime markers here. If the memory being set is still preserved, the lifetime intrinsics are correct.

DXIL does support lifetime markers, but we may not be correctly legalizing them.

Reply:
So comparing to DXC there should be no lifetime markers for this case. That said there isn’t an alloca either. I had thought something like mem2reg pass would clean up everything but it didn’t change the ir. I could try doing mem2reg after removing the memset and see that makes the lifetime markers go away and the alloca simplify.

Reply:
no mem2reg does not help here even after memset legalization. I'll remove lifetime legalization from this PR and create a new issue to track it.

Reply:
https://github.com/llvm/llvm-project/pull/136622